src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,739 @@
+    {
+        if (!IsValidOmrPrefix(onMeshPrefixConfig.GetPrefix()) || !onMeshPrefixConfig.mDefaultRoute ||
+            !onMeshPrefixConfig.mSlaac)
+        {
+            continue;
+        }
+
+        if (!IsValidOmrPrefix(lowestOmrPrefix) || onMeshPrefixConfig.GetPrefix() < lowestOmrPrefix)
+        {
+            lowestOmrPrefix = onMeshPrefixConfig.GetPrefix();
+        }
+    }
+
+    if (IsValidOmrPrefix(lowestOmrPrefix) && lowestOmrPrefix != mLocalOmrPrefix)
+    {
+        otLogInfoBr("EvaluateOmrPrefix: adopt existing OMR prefix %s in Thread network",
+                    lowestOmrPrefix.ToString().AsCString());
+        newOmrPrefix = lowestOmrPrefix;
+    }
+    else
+    {
+        newOmrPrefix = mLocalOmrPrefix;
+    }
+
+exit:
+    return newOmrPrefix;
+}
+
+void RoutingManager::PublishOmrPrefix(const Ip6::Prefix &aOmrPrefix)
+{
+    otError                         error;
+    NetworkData::OnMeshPrefixConfig omrPrefixConfig;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached());
+    VerifyOrExit(IsValidOmrPrefix(aOmrPrefix));
+
+    omrPrefixConfig.Clear();
+    omrPrefixConfig.mPrefix       = aOmrPrefix;
+    omrPrefixConfig.mStable       = true;
+    omrPrefixConfig.mSlaac        = true;
+    omrPrefixConfig.mPreferred    = true;
+    omrPrefixConfig.mOnMesh       = true;
+    omrPrefixConfig.mDefaultRoute = true;
+
+    error = Get<NetworkData::Local>().AddOnMeshPrefix(omrPrefixConfig);
+    if (error != OT_ERROR_NONE)
+    {
+        otLogInfoBr("failed to publish OMR prefix %s in Thread network: %s", aOmrPrefix.ToString().AsCString(),
+                    otThreadErrorToString(error));
+    }
+    else
+    {
+        Get<NetworkData::Notifier>().HandleServerDataUpdated();
+        otLogInfoBr("published OMR prefix %s in Thread network", aOmrPrefix.ToString().AsCString());
+    }
+
+exit:
+    return;
+}
+
+void RoutingManager::UnpublishOmrPrefix(const Ip6::Prefix &aOmrPrefix)
+{
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached());
+    VerifyOrExit(IsValidOmrPrefix(aOmrPrefix));
+
+    IgnoreError(Get<NetworkData::Local>().RemoveOnMeshPrefix(aOmrPrefix));
+    Get<NetworkData::Notifier>().HandleServerDataUpdated();
+
+exit:
+    return;
+}
+
+Ip6::Prefix RoutingManager::EvaluateOnLinkPrefix(void)
+{
+    Ip6::Prefix newOnLinkPrefix;
+
+    newOnLinkPrefix.Clear();
+
+    // We don't evalute on-link prefix if we are doing
+    // Router Discovery or we have already discovered some
+    // on-link prefixes.
+    VerifyOrExit(!mRouterSolicitTimer.IsRunning());
+    if (IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix))
+    {
+        otLogInfoBr("EvaluateOnLinkPrefix: there is already on-link prefix %s on interface %s",
+                    mDiscoveredOnLinkPrefix.ToString().AsCString(), GetInfraIfName());
+        ExitNow();
+    }
+
+    if (IsValidOnLinkPrefix(mAdvertisedOnLinkPrefix))
+    {
+        newOnLinkPrefix = mAdvertisedOnLinkPrefix;
+    }
+    else
+    {
+        newOnLinkPrefix = mLocalOnLinkPrefix;
+    }
```

Comment:
We should not rely on the non-trivial fact. I think the logic should be "if we have already been advertising some on-link prefix, then we keep advertising it. Otherwise, we advertise ourselves". And current code explains itself. thoughts?